transcriptomics/transcriptomics_matrix_generator.py

```diff
@@ -45,23 +45,32 @@ def load_sample_info(self, json_file_path):
             self.treatments = [item['Treatment'] for item in sample_info]
             self.experiment_id = [item['ExperimentID'] for item in sample_info]
 
-    def download_gff(self, url, file_path):   
+    def download_gff(self, url, file_path):
         if not os.path.exists(file_path):
             subprocess.run(['wget', '-O', file_path, url], check=True)
+        self.process_gff(file_path)
+
+    def process_gff(self, file_path):
         with gzip.open(file_path, 'rt') as file:
             gff_data = pd.read_csv(file, sep='\t', comment='#', header=None, names=[
                 'seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attribute'
             ], dtype=str)
         genes = gff_data[gff_data['feature'] == 'gene'].copy()
-        genes.loc[:, 'GeneName'] = genes['attribute'].str.extract('Name=([^;]+)')
+        genes.loc[:, 'GeneName'] = genes['attribute'].str.extract('Name=([^;]+)', expand=False)
         genes.loc[:, 'length'] = genes['end'].astype(int) - genes['start'].astype(int) + 1
-        genes = genes['GeneName'].dropna().tolist()
-        return genes
+        gene_info = genes[['GeneName', 'length']].dropna().drop_duplicates(subset='GeneName')
+        output_file_csv = 'gene_lengths.csv'
+        gene_info.to_csv(output_file_csv, index=False)
+        print(f"Gene lengths have been saved to {output_file_csv}.")
+
+        self.gene_names = gene_info['GeneName'].tolist()
+        self.num_genes = len(self.gene_names)
```

Comment:
In this kind of situation, when you need to hold a list of values (gene_names), and make decisions based the length of the list, it is best to simply get the length directly from the property when you need it. With `self.gene_names`, all we need to get the number of genes is `len(self.gene_names)`.

Otherwise, you have to maintain the state of 2 closely linked properties instead of just 1, which is not ideal for maintainability.